RecoBTag/DeepFlavour/plugins/DeepFlavourTagInfoProducer.cc

```diff
@@ -0,0 +1,318 @@
+	  ~DeepFlavourTagInfoProducer();
+
+	  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    typedef std::vector<reco::DeepFlavourTagInfo> DeepFlavourTagInfoCollection;
+    typedef reco::VertexCompositePtrCandidateCollection SVCollection;
+    typedef reco::VertexCollection VertexCollection;
+    typedef edm::View<reco::ShallowTagInfo> ShallowTagInfoCollection;
+
+	  virtual void beginStream(edm::StreamID) override {}
+	  virtual void produce(edm::Event&, const edm::EventSetup&) override;
+	  virtual void endStream() override {}
+
+    
+    const double jet_radius_;
+
+    edm::EDGetTokenT<edm::View<reco::Jet>>  jet_token_;
+    edm::EDGetTokenT<VertexCollection> vtx_token_;
+    edm::EDGetTokenT<SVCollection> sv_token_;
+    edm::EDGetTokenT<ShallowTagInfoCollection> shallow_tag_info_token_;
+    edm::EDGetTokenT<edm::ValueMap<float>> puppi_value_map_token_;
+    edm::EDGetTokenT<edm::ValueMap<int>> pvasq_value_map_token_;
+    edm::EDGetTokenT<edm::Association<VertexCollection>> pvas_token_;
+
+    bool use_puppi_value_map_;
+    bool use_pvasq_value_map_;
+    
+};
+
+DeepFlavourTagInfoProducer::DeepFlavourTagInfoProducer(const edm::ParameterSet& iConfig) :
+  jet_radius_(iConfig.getParameter<double>("jet_radius")),
+  jet_token_(consumes<edm::View<reco::Jet> >(iConfig.getParameter<edm::InputTag>("jets"))),
+  vtx_token_(consumes<VertexCollection>(iConfig.getParameter<edm::InputTag>("vertices"))),
+  sv_token_(consumes<SVCollection>(iConfig.getParameter<edm::InputTag>("secondary_vertices"))),
+  shallow_tag_info_token_(consumes<ShallowTagInfoCollection>(iConfig.getParameter<edm::InputTag>("shallow_tag_infos"))),
+  use_puppi_value_map_(false),
+  use_pvasq_value_map_(false)
+{
+  produces<DeepFlavourTagInfoCollection>();
+
+  const auto & puppi_value_map_tag = iConfig.getParameter<edm::InputTag>("puppi_value_map");
+  if (!puppi_value_map_tag.label().empty()) {
+    puppi_value_map_token_ = consumes<edm::ValueMap<float>>(puppi_value_map_tag);
+    use_puppi_value_map_ = true;
+  }
+
+  const auto & pvas_tag = iConfig.getParameter<edm::InputTag>("vertexAssociator");
+  if (!pvas_tag.label().empty()) {
+    pvasq_value_map_token_ = consumes<edm::ValueMap<int>>(pvas_tag);
+    pvas_token_ = consumes<edm::Association<VertexCollection>>(pvas_tag);
+    use_pvasq_value_map_ = true;
+  }
+
+}
+
+
+DeepFlavourTagInfoProducer::~DeepFlavourTagInfoProducer()
+{
+}
+
+void DeepFlavourTagInfoProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions)
+{
+}
+
+void DeepFlavourTagInfoProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+
+  auto output_tag_infos = std::make_unique<DeepFlavourTagInfoCollection>();
+
+  edm::Handle<edm::View<reco::Jet>> jets;
+  iEvent.getByToken(jet_token_, jets);
+
+  edm::Handle<VertexCollection> vtxs;
+  iEvent.getByToken(vtx_token_, vtxs);
+  // reference to primary vertex
+  const auto & pv = vtxs->at(0);
```

Comment:
some exception handling is needed here in case vtx collection is empty (there was a recent thread where it was actually a problem in a somewhat abnormal setup)

Reply:
will do